flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/asyncprocessing/AbstractAsyncStateStreamOperatorV2.java

```diff
@@ -188,6 +190,35 @@ public <K, N> InternalTimerService<N> getInternalTimerService(
                 (AsyncExecutionController<K>) asyncExecutionController);
     }
 
+    @Override
+    public void processWatermark(Watermark mark) throws Exception {
+        if (!isAsyncStateProcessingEnabled()) {
+            super.processWatermark(mark);
+            return;
+        }
+        asyncExecutionController.processNonRecord(() -> super.processWatermark(mark));
```

Comment:
We may need to override `processWatermark1` and `processWatermark2` as well, or we can override `processWatermark(Watermark, int)`, like that for `processWatermarkStatus`. I also understand it that this PR is mainly responsible for introducing the epoch mechanism, and that we would have another jira ticket and PR to apply epoch to all events and all cases. So it is also OK for me if you would like to make the change in the next PR.

Reply:
`AbstractStreamOperatorV2` does not implement `processWatermark1 `,  `processWatermark2` and `processWatermark(Watermark, int)`, if those methods will be added in the future, I would override them.